File: sic_backend_mvp_jwt_sqlite/app/schemas.py
from datetime import date, datetime, timedelta
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, EmailStr, Field, field_validator


class Token(BaseModel):
    access_token: str
    token_type: str = "bearer"


class UserCreate(BaseModel):
    email: EmailStr
    password: str = Field(min_length=8, max_length=128, strip_whitespace=True)
    referral_code: Optional[str] = Field(default=None, max_length=50, strip_whitespace=True)

    @field_validator('password')
    @classmethod
    def validate_password_strength(cls, v: str) -> str:
        if len(v) < 8:
            raise ValueError('Password must be at least 8 characters long')
        if not any(c.isupper() for c in v):
            raise ValueError('Password must contain at least one uppercase letter')
        if not any(c.islower() for c in v):
            raise ValueError('Password must contain at least one lowercase letter')
        if not any(c.isdigit() for c in v):
            raise ValueError('Password must contain at least one digit')
        return v


class UserOut(BaseModel):
    id: int
    email: EmailStr
    model_config = ConfigDict(from_attributes=True)


class ClientCreate(BaseModel):
    name: str = Field(min_length=1, max_length=255, strip_whitespace=True)
    email: EmailStr
    contact_name: Optional[str] = Field(default=None, max_length=255, strip_whitespace=True)
    contact_phone: Optional[str] = Field(default=None, max_length=20, strip_whitespace=True)
    timezone: str = Field(default="UTC", max_length=64)
    notes: Optional[str] = Field(default=None, max_length=1000)

    @field_validator('timezone')
    @classmethod
    def validate_timezone(cls, v: str) -> str:
        try:
            from zoneinfo import ZoneInfo
            ZoneInfo(v)
            return v
        except Exception:
            return "UTC"  # Fallback to UTC for invalid timezones


class ClientUpdate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=255, strip_whitespace=True)
    email: Optional[EmailStr] = None
    contact_name: Optional[str] = Field(default=None, max_length=255, strip_whitespace=True)
    contact_phone: Optional[str] = Field(default=None, max_length=20, strip_whitespace=True)
    timezone: Optional[str] = Field(default=None, max_length=64)
    notes: Optional[str] = Field(default=None, max_length=1000)

    @field_validator('timezone')
    @classmethod
    def validate_timezone(cls, v: Optional[str]) -> Optional[str]:
        if v is None:
            return v
        try:
            from zoneinfo import ZoneInfo
            ZoneInfo(v)
            return v
        except Exception:
            return "UTC"  # Fallback to UTC for invalid timezones


class ClientOut(BaseModel):
    id: int
    name: str
    email: EmailStr
    contact_name: Optional[str] = None
    contact_phone: Optional[str] = None
    timezone: str
    notes: Optional[str] = None
    model_config = ConfigDict(from_attributes=True)


class InvoiceCreate(BaseModel):
    client_id: int = Field(gt=0)
    amount_cents: int = Field(gt=0, le=999999999)  # Max $9,999,999.99
    currency: str = Field(default="USD", strip_whitespace=True)
    due_date: date
    status: Literal["draft", "pending", "paid", "overdue", "cancelled"] = "pending"
    external_id: Optional[str] = Field(default=None, max_length=255, strip_whitespace=True)
    source: str = Field(default="manual", max_length=32)

    @field_validator('currency')
    @classmethod
    def validate_currency(cls, v: str) -> str:
        # Strip and uppercase, ensure exactly 3 characters
        v_clean = v.strip().upper() if v else "USD"
        if len(v_clean) != 3:
            return "USD"  # Default to USD for invalid length

        # Only allow common currency codes
        allowed_currencies = {"USD", "EUR", "GBP", "CAD", "AUD", "JPY", "CHF"}
        if v_clean not in allowed_currencies:
            return "USD"  # Default to USD for invalid currencies
        return v_clean

    @field_validator('due_date')
    @classmethod
    def validate_due_date(cls, v: date) -> date:
        from datetime import date as date_type
        today = date_type.today()
        if v < today:
            raise ValueError('Due date cannot be in the past')
        # Reasonable limit - max 5 years in the future
        max_future = today.replace(year=today.year + 5)
        if v > max_future:
            raise ValueError('Due date cannot be more than 5 years in the future')
        return v


class InvoiceUpdate(BaseModel):
    client_id: Optional[int] = None
    amount_cents: Optional[int] = Field(default=None, gt=0)
    currency: Optional[str] = None
    due_date: Optional[date] = None
    status: Optional[Literal["draft", "pending", "paid", "overdue", "cancelled"]] = None
    external_id: Optional[str] = None
    source: Optional[str] = None

    @field_validator("currency")
    @classmethod
    def normalize_currency_opt(cls, v: Optional[str]) -> Optional[str]:
        if v is None:
            return v
        v2 = (v or "").strip().upper()
        if len(v2) != 3:
            raise ValueError("currency must be a 3-letter code")
        return v2


class InvoiceOut(BaseModel):
    id: int
    client_id: int
    amount_cents: int
    currency: str
    due_date: date
    status: str
    external_id: Optional[str] = None
    source: str
    payment_link_url: Optional[str] = None
    paid_at: Optional[datetime] = None
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


class ReminderCreate(BaseModel):
    invoice_id: int = Field(gt=0)
    send_at: datetime
    channel: Literal["email", "sms", "whatsapp"] = "email"
    subject: Optional[str] = Field(default=None, max_length=255, strip_whitespace=True)
    body: Optional[str] = Field(default=None, max_length=5000, strip_whitespace=True)

    @field_validator('send_at')
    @classmethod
    def validate_send_at(cls, v: datetime) -> datetime:
        from datetime import datetime, timezone
        now = datetime.now(timezone.utc)
        if v < now - timedelta(minutes=5):
            raise ValueError('Reminder send time must be within the last 5 minutes or the future')
        # Limit to reasonable future (1 year)
        max_future = now.replace(year=now.year + 1)
        if v > max_future:
            raise ValueError('Reminder send time cannot be more than 1 year in the future')
        return v


class ReminderUpdate(BaseModel):
    send_at: Optional[datetime] = None
    channel: Optional[Literal["email", "sms", "whatsapp"]] = None
    status: Optional[Literal["scheduled", "sent", "failed", "cancelled"]] = None
    subject: Optional[str] = None
    body: Optional[str] = None


class ReminderOut(BaseModel):
    id: int
    invoice_id: int
    send_at: datetime
    channel: str
    status: str
    subject: Optional[str] = None
    body: Optional[str] = None
    created_at: Optional[datetime] = None
    model_config = ConfigDict(from_attributes=True)


# Templates
class TemplateCreate(BaseModel):
    name: str
    tone: Literal["friendly", "neutral", "firm"]
    subject: str
    body_markdown: str


class TemplateUpdate(BaseModel):
    name: Optional[str] = None
    tone: Optional[Literal["friendly", "neutral", "firm"]] = None
    subject: Optional[str] = None
    body_markdown: Optional[str] = None


class TemplateOut(BaseModel):
    id: int
    user_id: int
    name: str
    tone: str
    subject: str
    body_markdown: str
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


# Events
class EventOut(BaseModel):
    id: int
    user_id: int
    entity_type: str
    entity_id: int
    event_type: str
    payload: dict
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


class ReminderPreviewRequest(BaseModel):
    # Legacy path (uses invoice to derive variables)
    invoice_id: Optional[int] = None
    template_id: Optional[int] = None
    tone: Optional[Literal["friendly", "neutral", "firm"]] = None
    # New path (AC): explicit template name and variables
    template: Optional[str] = None
    variables: Optional[dict] = None


class ReminderPreviewOut(BaseModel):
    subject: str
    html: str
    text: str


class ReminderSendNowRequest(BaseModel):
    reminder_id: int
    force: bool = False


class ReminderSendNowOut(BaseModel):
    id: int
    status: str
    meta: dict | None = None


class DeadLetterOut(BaseModel):
    id: int
    kind: str
    payload: dict
    error: str
    retries: int
    next_attempt_at: Optional[datetime] = None
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


# Outbox
class OutboxOut(BaseModel):
    id: int
    topic: str
    payload: dict
    status: str
    attempts: int
    next_attempt_at: Optional[datetime] = None
    dispatched_at: Optional[datetime] = None
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


# Analytics
class AnalyticsStatusTotals(BaseModel):
    all: int
    draft: int
    pending: int
    paid: int
    overdue: int
    cancelled: int


class AnalyticsTopLateClient(BaseModel):
    client_id: int
    client_name: str
    client_email: str
    avg_days_late: float
    overdue_count: int
    total_overdue_amount_cents: int


class CurrencyRevenueBreakdown(BaseModel):
    currency: str
    earned_revenue: int
    outstanding_revenue: int
    total_revenue: int


class AnalyticsSummaryOut(BaseModel):
    totals: AnalyticsStatusTotals
    earned_revenue: int  # Total revenue from paid invoices (in cents)
    expected_payments_next_30d: int
    avg_days_to_pay: Optional[float] = None
    top_late_clients: list[AnalyticsTopLateClient]
    currency_breakdown: Optional[list[CurrencyRevenueBreakdown]] = None


class AnalyticsTimeseriesPoint(BaseModel):
    period: str  # ISO date string for the period start
    value: float
    count: Optional[int] = None  # Number of items in this period


class AnalyticsTimeseriesOut(BaseModel):
    metric: str
    interval: str
    points: list[AnalyticsTimeseriesPoint]
    total_value: float
    total_count: int


# Events
class EventCreate(BaseModel):
    entity_type: str
    entity_id: int
    event_type: str
    payload: Optional[dict] = None


class EventOut(BaseModel):
    id: int
    user_id: int
    entity_type: str
    entity_id: int
    event_type: str
    payload: Optional[dict] = None
    created_at: datetime
    updated_at: datetime
    model_config = ConfigDict(from_attributes=True)

File: sic_app/src/lib/schemas.ts
import { z } from 'zod'

// Auth schemas
export const loginSchema = z.object({
  email: z
    .string()
    .min(1, 'Email is required')
    .email('Please enter a valid email address'),
  password: z
    .string()
    .min(1, 'Password is required')
    .min(6, 'Password must be at least 6 characters'),
})

export const registerSchema = z
  .object({
    email: z
      .string()
      .min(1, 'Email is required')
      .email('Please enter a valid email address'),
    password: z
      .string()
      .min(6, 'Password must be at least 6 characters')
      .max(128, 'Password must be less than 128 characters'),
    confirmPassword: z
      .string()
      .min(1, 'Please confirm your password'),
    referralCode: z
      .string()
      .optional()
      .refine((value) => !value || (value.length === 8 && /^[A-Z0-9]+$/.test(value)), {
        message: "Referral code must be 8 characters long and contain only letters and numbers"
      }),
  })
  .refine((data) => data.password === data.confirmPassword, {
    message: "Passwords don't match",
    path: ['confirmPassword'],
  })

export type LoginFormData = z.infer<typeof loginSchema>
export type RegisterFormData = z.infer<typeof registerSchema>

// Client schemas
export const clientSchema = z.object({
  name: z
    .string()
    .min(1, 'Client name is required')
    .max(255, 'Name must be less than 255 characters'),
  email: z
    .string()
    .min(1, 'Email is required')
    .email('Please enter a valid email address')
    .max(255, 'Email must be less than 255 characters'),
  timezone: z
    .string()
    .optional()
    .default('UTC'),
})

export type ClientFormData = z.infer<typeof clientSchema>

// Invoice schemas
export const invoiceSchema = z.object({
  client_id: z
    .number({
      required_error: 'Please select a client',
      invalid_type_error: 'Please select a client',
    })
    .positive('Please select a client'),
  amount_cents: z
    .number({
      required_error: 'Amount is required',
      invalid_type_error: 'Amount must be a number',
    })
    .positive('Amount must be greater than 0')
    .max(999999999, 'Amount is too large'), // $9,999,999.99
  currency: z
    .string()
    .min(1, 'Currency is required')
    .length(3, 'Currency must be 3 characters')
    .regex(/^[A-Z]{3}$/, 'Currency must be 3 uppercase letters'),
  due_date: z
    .string()
    .min(1, 'Due date is required')
    .refine(
      (date) => {
        const parsed = new Date(date)
        return !isNaN(parsed.getTime())
      },
      { message: 'Please enter a valid date' }
    )
    .refine(
      (date) => {
        const parsed = new Date(date)
        const today = new Date()
        today.setHours(0, 0, 0, 0)
        return parsed >= today
      },
      { message: 'Due date cannot be in the past' }
    ),
  status: z
    .enum(['draft', 'pending', 'paid', 'overdue', 'cancelled'])
    .default('pending'),
})

export const updateInvoiceSchema = z.object({
  status: z
    .enum(['draft', 'pending', 'paid', 'overdue', 'cancelled'])
    .optional(),
  paid_at: z
    .string()
    .nullable()
    .optional()
    .refine(
      (date) => {
        if (!date) return true
        const parsed = new Date(date)
        return !isNaN(parsed.getTime())
      },
      { message: 'Please enter a valid date and time' }
    ),
})

export type InvoiceFormData = z.infer<typeof invoiceSchema>
export type UpdateInvoiceFormData = z.infer<typeof updateInvoiceSchema>

// Reminder schemas  
export const reminderSchema = z.object({
  invoice_id: z
    .number({
      required_error: 'Please select an invoice',
      invalid_type_error: 'Please select an invoice',
    })
    .positive('Please select an invoice'),
  send_at: z
    .string()
    .min(1, 'Send date is required')
    .refine(
      (date) => {
        const parsed = new Date(date)
        return !isNaN(parsed.getTime())
      },
      { message: 'Please enter a valid date and time' }
    )
    .refine(
      (date) => {
        const parsed = new Date(date)
        const now = new Date()
        return parsed > now
      },
      { message: 'Send date must be in the future' }
    ),
  channel: z
    .enum(['email', 'sms', 'whatsapp'])
    .default('email'),
  subject: z
    .string()
    .max(255, 'Subject must be less than 255 characters')
    .optional(),
  body: z
    .string()
    .max(2000, 'Message must be less than 2000 characters')
    .optional(),
})

export type ReminderFormData = z.infer<typeof reminderSchema>

// Template schemas
export const templateSchema = z.object({
  name: z
    .string()
    .min(1, 'Template name is required')
    .max(255, 'Name must be less than 255 characters'),
  tone: z
    .enum(['friendly', 'neutral', 'firm'])
    .default('friendly'),
  subject: z
    .string()
    .min(1, 'Subject is required')
    .max(255, 'Subject must be less than 255 characters'),
  body_markdown: z
    .string()
    .min(1, 'Template body is required')
    .max(5000, 'Template body must be less than 5000 characters'),
})

export type TemplateFormData = z.infer<typeof templateSchema>

// Search and filter schemas
export const invoiceFiltersSchema = z.object({
  client_id: z.number().optional(),
  status: z.string().optional(),
  search: z.string().optional(),
  from_date: z.string().optional(),
  to_date: z.string().optional(),
})

export const clientFiltersSchema = z.object({
  search: z.string().optional(),
})

export type InvoiceFilters = z.infer<typeof invoiceFiltersSchema>
export type ClientFilters = z.infer<typeof clientFiltersSchema>

// Settings schemas
export const profileSchema = z.object({
  email: z
    .string()
    .min(1, 'Email is required')
    .email('Please enter a valid email address'),
  current_password: z
    .string()
    .optional(),
  new_password: z
    .string()
    .min(6, 'Password must be at least 6 characters')
    .optional()
    .or(z.literal('')),
  confirm_password: z
    .string()
    .optional()
    .or(z.literal('')),
})
.refine(
  (data) => {
    if (data.new_password && !data.confirm_password) {
      return false
    }
    if (data.new_password !== data.confirm_password) {
      return false
    }
    return true
  },
  {
    message: "Passwords don't match",
    path: ['confirm_password'],
  }
)
.refine(
  (data) => {
    if (data.new_password && !data.current_password) {
      return false
    }
    return true
  },
  {
    message: 'Current password is required to set a new password',
    path: ['current_password'],
  }
)

export type ProfileFormData = z.infer<typeof profileSchema>

// Validation helpers
export function validateCurrency(currency: string): boolean {
  return /^[A-Z]{3}$/.test(currency)
}

export function validateAmount(amount: string): number | null {
  const num = parseFloat(amount)
  if (isNaN(num) || num <= 0) return null
  return Math.round(num * 100) // Convert to cents
}

export function formatAmountForInput(amountCents: number): string {
  return (amountCents / 100).toFixed(2)
}